check the dts file for mtd name, the auto generated name may not be compatible with uboot settings
check the dts file for presence of bbt-use-flash variable, use bad block table in the flash (by default OOB is used)
http://lists.infradead.org/pipermail/linux-mtd/2013-March/046371.html

--- a/drivers/mtd/nand/plat_nand.c
+++ b/drivers/mtd/nand/plat_nand.c
@@ -12,6 +12,7 @@
 #include <linux/io.h>
 #include <linux/module.h>
 #include <linux/platform_device.h>
+#include <linux/of.h>
 #include <linux/slab.h>
 #include <linux/mtd/mtd.h>
 #include <linux/mtd/nand.h>
@@ -35,6 +36,7 @@ static int plat_nand_probe(struct platfo
 	struct plat_nand_data *data;
 	struct resource *res;
 	const char **part_types;
+	const char __maybe_unused	*of_mtd_name = NULL;
 	int err = 0;
 
 	if (!pdata) {
@@ -75,7 +77,20 @@ static int plat_nand_probe(struct platfo
 	data->chip.priv = &data;
 	data->mtd.priv = &data->chip;
 	data->mtd.owner = THIS_MODULE;
-	data->mtd.name = dev_name(&pdev->dev);
+	if (pdev->dev.of_node) {
+		int i;
+		of_property_read_string(pdev->dev.of_node,
+						"linux,mtd-name", &of_mtd_name);
+		if (of_mtd_name)
+			data->mtd.name = of_mtd_name;
+		else
+			data->mtd.name = dev_name(&pdev->dev);
+		if (of_get_property(pdev->dev.of_node, "bbt-use-flash", &i))
+			data->chip.bbt_options |= NAND_BBT_USE_FLASH;
+	} else {
+		data->mtd.name = dev_name(&pdev->dev);
+		data->chip.bbt_options |= pdata->chip.bbt_options;
+	}
 
 	data->chip.IO_ADDR_R = data->io_base;
 	data->chip.IO_ADDR_W = data->io_base;
@@ -87,7 +102,6 @@ static int plat_nand_probe(struct platfo
 	data->chip.read_byte = pdata->ctrl.read_byte;
 	data->chip.chip_delay = pdata->chip.chip_delay;
 	data->chip.options |= pdata->chip.options;
-	data->chip.bbt_options |= pdata->chip.bbt_options;
 
 	data->chip.ecc.hwctl = pdata->ctrl.hwcontrol;
 	data->chip.ecc.layout = pdata->chip.ecclayout;
